### PR TITLE
Better organise the documentation, document the policy engine, the authorisation process and the scopes

### LINKED ISSUE #2187: Document the default policy behaviour around admin scopes
The default policy is not documented at all, but it is especially tricky to understand how admin scopes (`urn:synapse:admin:*` and `urn:mas:admin`) interact with it. This needs to be documented

A few informations to include, in a very brain-dumpy format:

 - There are three main ways to get an access token:
   - Interactively, through an authorisation code grant, in which case the token has a user attached to it. *This is regulated by the policy engine*
   - Programatically, through a client credentials grant, in which case the token has no user attached to it, only an OAuth client. *This is regulated by the policy engine*
   - Programatically, through MAS' GraphQL API, using the `createOauth2Session` mutation. This is completely unregulated, arbitrary (including non-existing) scopes can be granted through this. The only restriction is that the token making the API call needs to be admin (= has the `urn:mas:admin` scope)
 - The `urn:synapse:admin:*` scope grants access to `/_synapse/admin/*`. Because of some Synapse subtlety, this scope has to be linked to a user, so it can't be requested through the client credentials grant
 - The `urn:mas:admin` scope grants admin access to MAS' GraphQL API, which lets it request and edit resources of other users than the current one, and access a few privileged mutations. When requesting this scope, the `urn:mas:graphql:*` scope should be requested as well.
 - With the default policy, a user [can request the `urn:synapse:admin:*` scope](https://github.com/matrix-org/matrix-authentication-service/blob/bd42916309d3237f1523066a6238d520c4cd6f73/policies/authorization_grant.rego#L34-L38) or the [`urn:mas:admin` scope](https://github.com/matrix-org/matrix-authentication-service/blob/bd42916309d3237f1523066a6238d520c4cd6f73/policies/authorization_grant.rego#L44-L47) if:
   - [their *username* (aka. localpart of the MXID) is listed in `policy.data.admin_users`](https://github.com/matrix-org/matrix-authentication-service/blob/bd42916309d3237f1523066a6238d520c4cd6f73/policies/authorization_grant.rego#L15-L19)
   - or if [their account has the `can_request_admin` flag in the database set to true](https://github.com/matrix-org/matrix-authentication-service/blob/bd42916309d3237f1523066a6238d520c4cd6f73/policies/authorization_grant.rego#L22-L24) (see below about toggling that flag)
 - With the default policy, [a client can request the `urn:mas:admin` scope if their client ID is listed in `policy.data.admin_clients`](https://github.com/matrix-org/matrix-authentication-service/blob/bd42916309d3237f1523066a6238d520c4cd6f73/policies/authorization_grant.rego#L50-L54)
 - The [GraphQL schema is self-documented](https://github.com/matrix-org/matrix-authentication-service/blob/bd42916309d3237f1523066a6238d520c4cd6f73/frontend/schema.graphql). A playground can be found in MAS itself, for example: https://auth-oidc.lab.element.dev/graphql/playground
 - The URL of the GraphQL endpoint can be discovered with the `org.matrix.matrix-authentication-service.graphql_endpoint` property [in the OIDC discovery document](https://auth-oidc.lab.element.dev/.well-known/openid-configuration)
 - The `can_request_admin` flag can be found on the `users` table in the database, and can be dynamically toggled using the [`setCanRequestAdmin` mutation](https://github.com/matrix-org/matrix-authentication-service/blob/bd42916309d3237f1523066a6238d520c4cd6f73/frontend/schema.graphql#L665-L671). This mutation [requires a `userId`](https://github.com/matrix-org/matrix-authentication-service/blob/bd42916309d3237f1523066a6238d520c4cd6f73/frontend/schema.graphql#L1054-L1066), which can be found on users using the [`userByUsername`](https://github.com/matrix-org/matrix-authentication-service/blob/bd42916309d3237f1523066a6238d520c4cd6f73/frontend/schema.graphql#L888-L892) query method

### PR DESCRIPTION
**Pull request by sandhose**
This adds a few section to the documentation:

 - [How the policy engine works](https://quenting-api-docs.matrix-authentication-service-docs.pages.dev/topics/policy)
 - [How sessions and authorisations work](https://quenting-api-docs.matrix-authentication-service-docs.pages.dev/topics/authorization)
 - [Reference documentation on the scopes](https://quenting-api-docs.matrix-authentication-service-docs.pages.dev/reference/scopes)
 - [Reference documentation on the GraphQL API](https://quenting-api-docs.matrix-authentication-service-docs.pages.dev/reference/graphql)

Fixes #2187
Part of #2605

This also reorganises the documentation to have basically 4 parts:

 - A setup guide
 - Big "topics" high level explanations
 - Reference documentation
 - Technical documentation

I also want to add more "how-to" tutorials, really task-oriented, but I'll add some of them in another PR.

### CODE REVIEW

File: docs/reference/scopes.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+This allows the authenticated entity to perform any operation on the API, regardless of whether the entity owns the data or not.
+
+The default policy doesn't allow everyone to request this scope.
+It allows:
+
+- for the "[authorization code]" and "[device authorization]" grants:
+  - users with the `can_request_admin` attribute set to `true` in the database
+  - users listed in the [`policy.data.admin_users`](../reference/configuration.md#policy) configuration option
+- for the "client credentials" grant:
+  - clients that are listed in the [`policy.data.admin_clients`](../reference/configuration.md#policy) configuration option
```
Comment (gabreal) [inline docs/reference/scopes.md:103]: `admin_clients` are not documented there yet only `admin_users`. are clients configured in the configuration file automatically added to the `admin_clients` list?

File: docs/reference/scopes.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+## MAS-specific scopes
+
+MAS also has a few scopes that are specific to the MAS implementation.
+
+### `urn:mas:graphql:*`
+
+This scope grants access to the whole MAS [GraphQL API].
+What permission the session has on the API is determined by the entity that the session is authorized as.
+If authorized as a user, this will usually allow querying and mutating the user's data.
```
Comment (gabreal) [inline docs/reference/scopes.md:86]: this is a bit unclear to me: is *authorized as a user* equal to not having authorized requesting the `urn:mas:admin` scope?

**Review by gabreal — COMMENTED** [review]
thanks a lot for adding this detailed documentation

File: docs/reference/scopes.md (hunk 1 above)
Comment (sandhose) [reply docs/reference/scopes.md:103]: You're right! I forgot to document this, I added this in 31a8add

File: docs/reference/scopes.md (hunk 2 above)
Comment (sandhose) [reply docs/reference/scopes.md]: This was indeed not well phrased. I tried to clarify this in 83563ea3, what do you think?

Comment (gabreal) [reply docs/reference/scopes.md]: that makes it clear. thank you!

File: docs/topics/authorization.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+In MAS, the access token is an opaque string for which the service has metadata associated with it.
+An access token has:
+
+- a subject, which is the user the token is issued for
+- a list of [scopes](../reference/scopes.md)
+- a client for which the token is issued
+- a timeframe for which the token is valid
+
+On a single token, those metadata are immutable: they don't change over time.
```
Comment (wrjlewis) [inline docs/topics/authorization.md:16]:
```suggestion
On a single token, metadata is immutable: they don't change over time.
```

File: docs/topics/authorization.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+- the `active` field, which tells if the token is valid or not
+- the `sub` field, which tells which user the token is issued for. This is an opaque string, and Synapse saves the mapping between the Matrix user ID and the subject of the token in its own database
+- in case Synapse doesn't know the presented subject, it will look at the `username` field, which it will use as the localpart for the user as fallback
+- the `scope` field, which tells which scopes are granted to the token. More specifically, it will look for the following scopes:
+  - [`urn:matrix:org.matrix.msc2967.client:api:*`], which grants broad access to the whole Matrix C-S API
+  - [`urn:matrix:org.matrix.msc2967.client:device:AABBCC`], which encodes the Matrix device ID used by the client
+  - [`urn:synapse:admin:*`], which grants access to the Synapse admin API
+
+It's important to understand that when Synapse delegates the authentication to MAS, a lot of user attributes are not managed by Synapse anymore.
```
Comment (wrjlewis) [inline docs/topics/authorization.md:37]:
```suggestion
It's important to understand that when Synapse delegates authentication to MAS, Synapse no longer manages many user attributes.
```

File: docs/topics/authorization.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+One important missing scope is [`urn:synapse:admin:*`], which means that the client won't have access to the Synapse admin API.
+
+This is the case even if the user has the `can_request_admin` attribute set to `true`, and this is by design:
+the legacy login API doesn't have a way to request specific scopes, and we don't want to grant admin access to all clients that have a compatibility session.
+This was the case in the past with Synapse, as the admin status was set on the user itself, but this is not the case anymore with MAS.
+
+## OAuth 2.0 sessions
+
+Modern clients are expected to use OAuth 2.0 to authenticate with the homeserver.
+In OAuth 2.0/OIDC, there are multiple way to start an OAuth 2.0 session called grants.
```
Comment (wrjlewis) [inline docs/topics/authorization.md:60]:
```suggestion
In OAuth 2.0/OIDC, there are multiple ways to start an OAuth 2.0 session called grants.
```

File: docs/topics/policy.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+The policy is evaluated in three different scenarios:
+
+ - [`register.rego`]: During user registration, either with password credentials or with an upstream OAuth 2.0 provider. This calls the [`email.rego`] and [`password.rego`] policies as well.
+ - [`email.rego`]: When a user adds a new email address to their account.
+ - [`password.rego`]: When a user changes their password.
+
+### Client registration
+
+The policy ([`client_registration.rego`]) is evaluated when a client send their metadata through the OAuth 2.0 dynamic client registration API.
```
Comment (wrjlewis) [inline docs/topics/policy.md:35]:
```suggestion
The policy ([`client_registration.rego`]) is evaluated when a client sends their metadata through the OAuth 2.0 dynamic client registration API.
```

**Review by wrjlewis — COMMENTED** [review]
Looks great! Nicely organised documentation and easy to read.

File: docs/SUMMARY.md [hunk 7]
```diff
@@ -17,17 +17,23 @@
...
+- [OAuth 2.0 scopes](./reference/scopes.md)
+- [Command line tool](./reference/cli/README.md)
+    - [`config`](./reference/cli/config.md)
+    - [`database`](./reference/cli/database.md)
+    - [`manage`](./reference/cli/manage.md)
+    - [`server`](./reference/cli/server.md)
+    - [`templates`](./reference/cli/templates.md)
+    - [`doctor`](./reference/cli/doctor.md)
 
 # Development
```
[the remark this answers is not in the document]
Comment (sandhose) [reply docs/SUMMARY.md:39]: It is indeed a bit lonely, its main purpose is to have a link from the README to have a disclaimer on the state of ASes. I think I'll leave it as-is for now

**Pull request merged**